heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```diff
@@ -139,107 +142,129 @@ public HeronMasterDriver(EvaluatorRequestor requestor,
-      throws InterruptedException {
-    executor.submit(new Callable<AllocatedEvaluator>() {
+  @VisibleForTesting
+  AllocatedEvaluator launchContainerForExecutor(final String executorId,
+                                                        final int cpu,
+                                                        final int mem) {
+    return launchContainerForExecutor(new HeronWorker(executorId, cpu, mem));
+  }
+
+  private AllocatedEvaluator launchContainerForExecutor(final HeronWorker worker) {
+    Future<AllocatedEvaluator> result = executor.submit(new Callable<AllocatedEvaluator>() {
       @Override
       public AllocatedEvaluator call() throws Exception {
-        AllocatedEvaluator evaluator = allocateContainer(executorId, cpu, mem);
+        AllocatedEvaluator evaluator = allocateContainer(worker.workerId, worker.cores, worker.mem);
+        if (evaluator == null) {
+          LOG.log(Level.WARNING, "Could not allocate container for id {0}", worker.workerId);
+          return null;
+        }
 
-        Configuration context = createContextConfig(executorId);
+        Configuration context = createContextConfig(worker.workerId);
 
         LOG.log(Level.INFO,
             "Activating container {0} for heron executor, id: {1}",
-            new Object[]{evaluator.getId(), executorId});
+            new Object[]{evaluator.getId(), worker.workerId});
         evaluator.submitContext(context);
-        reefContainerToHeronExecutorMap.put(evaluator.getId(), executorId);
+        multiKeyWorkerMap.assignEvaluatorToWorker(worker, evaluator);
         return evaluator;
       }
     });
+
+    try {
+      return result.get();
+    } catch (InterruptedException e) {
+      LOG.log(Level.SEVERE, "Interrupted while waiting for container", e);
+    } catch (ExecutionException e) {
+      LOG.log(Level.SEVERE, "Error while waiting for container", e);
+    }
+    return null;
   }
 
-  AllocatedEvaluator allocateContainer(String id, int cpu, int mem) throws InterruptedException {
+  AllocatedEvaluator allocateContainer(String id, int cpu, int mem) {
     EvaluatorRequest evaluatorRequest = createEvaluatorRequest(cpu, mem);
 
     LOG.log(Level.INFO, "Requesting container for executor, id: {0}, mem: {1}, cpu: {2}",
         new Object[]{id, mem, cpu});
     requestor.submit(evaluatorRequest);
 
-    AllocatedEvaluator evaluator = allocatedContainerQ.take();
-    LOG.log(Level.INFO,
-        "Container {0} is allocated for executor, id: {1}",
-        new Object[]{evaluator.getId(), id});
-    return evaluator;
+    try {
+      AllocatedEvaluator evaluator = allocatedContainerQ.take();
+      LOG.log(Level.INFO,
+          "Container {0} is allocated for executor, id: {1}",
+          new Object[]{evaluator.getId(), id});
+      return evaluator;
+    } catch (InterruptedException e) {
+      LOG.log(Level.WARNING, "Interrupted while waiting for container allocation", e);
+      return null;
```

Comment:
this seems to be an exceptional case. can we throw an exception instead of returning null?